File: legacy/board-express/backend/src/routes/sprints.ts
/**
 * Sprint routes.
 *
 *   GET    /api/sprints                    -> list (with per-sprint rollups)
 *   POST   /api/sprints                    -> create
 *   GET    /api/sprints/:id                -> detail (with assigned cards)
 *   PATCH  /api/sprints/:id                -> edit any subset of fields
 *   POST   /api/sprints/:id/cards          -> assign / re-assign a card
 *   DELETE /api/sprints/:id/cards/:cardId  -> remove a card from a sprint
 *
 * Wire format is camelCase end-to-end. The SQL columns are snake_case,
 * but we shape on the way out and validate on the way in so the
 * frontend never sees the database's spellings.
 */

import { Router, type Request, type Response } from "express";

import { getDb } from "../db/sqlite.js";

export const SPRINT_STATUSES = [
  "planning",
  "active",
  "completed",
  "cancelled",
] as const;
export type SprintStatus = (typeof SPRINT_STATUSES)[number];

const VALID_STATUSES = new Set<string>(SPRINT_STATUSES);

interface SprintRow {
  id: number;
  name: string;
  starts_at: string;
  ends_at: string;
  goal: string | null;
  status: string;
  points_target: number | null;
  dollar_target: number | null;
  review_hours_target: number | null;
  archived_at: string | null;
  created_at: string;
}

interface SprintCardRow {
  sprint_id: number;
  card_id: string;
  planned_points: number | null;
}

interface SprintRollupRow {
  sprint_id: number;
  card_count: number;
  points_sum: number | null;
}

export interface SprintShape {
  id: number;
  name: string;
  startsAt: string;
  endsAt: string;
  goal: string | null;
  status: string;
  pointsTarget: number | null;
  dollarTarget: number | null;
  reviewHoursTarget: number | null;
  archivedAt: string | null;
  createdAt: string;
}

export interface SprintSummary extends SprintShape {
  cardCount: number;
  plannedPointsSum: number;
}

function shapeSprint(row: SprintRow): SprintShape {
  return {
    id: row.id,
    name: row.name,
    startsAt: row.starts_at,
    endsAt: row.ends_at,
    goal: row.goal,
    status: row.status,
    pointsTarget: row.points_target,
    dollarTarget: row.dollar_target,
    reviewHoursTarget: row.review_hours_target,
    archivedAt: row.archived_at,
    createdAt: row.created_at,
  };
}

const SELECT_COLUMNS =
  "id, name, starts_at, ends_at, goal, status, points_target, dollar_target, review_hours_target, archived_at, created_at";

export function sprintsRouter(): Router {
  const router = Router();

  router.get("/sprints", (req: Request, res: Response) => {
    const includeArchived =
      typeof req.query["includeArchived"] === "string" &&
      ["1", "true", "yes"].includes(
        String(req.query["includeArchived"]).toLowerCase()
      );

    const sprints = (
      getDb()
        .prepare(
          `SELECT ${SELECT_COLUMNS} FROM sprints ${
            includeArchived ? "" : "WHERE archived_at IS NULL"
          } ORDER BY starts_at DESC`
        )
        .all() as SprintRow[]
    ).map(shapeSprint);

    if (sprints.length === 0) {
      res.json({ sprints: [] });
      return;
    }

    const rollups = getDb()
      .prepare(
        `SELECT sprint_id, COUNT(*) AS card_count, COALESCE(SUM(planned_points), 0) AS points_sum
           FROM sprint_cards
          GROUP BY sprint_id`
      )
      .all() as SprintRollupRow[];

    const rollupById = new Map<number, SprintRollupRow>();
    for (const r of rollups) rollupById.set(r.sprint_id, r);

    const summaries: SprintSummary[] = sprints.map((s) => {
      const r = rollupById.get(s.id);
      return {
        ...s,
        cardCount: r?.card_count ?? 0,
        plannedPointsSum: r?.points_sum ?? 0,
      };
    });

    res.json({ sprints: summaries });
  });

  router.post("/sprints", (req: Request, res: Response) => {
    const body = req.body as
      | {
          name?: unknown;
          startsAt?: unknown;
          endsAt?: unknown;
          goal?: unknown;
          status?: unknown;
        }
      | undefined;
    const name = typeof body?.name === "string" ? body.name.trim() : "";
    const startsAt = typeof body?.startsAt === "string" ? body.startsAt : "";
    const endsAt = typeof body?.endsAt === "string" ? body.endsAt : "";
    const goal =
      typeof body?.goal === "string" && body.goal.length > 0
        ? body.goal
        : null;
    const status =
      typeof body?.status === "string" && VALID_STATUSES.has(body.status)
        ? body.status
        : "planning";

    if (!name) {
      res.status(400).json({ error: "name is required" });
      return;
    }
    if (!startsAt || !endsAt) {
      res
        .status(400)
        .json({ error: "startsAt and endsAt are required (ISO date)" });
      return;
    }
    if (endsAt < startsAt) {
      res.status(400).json({ error: "endsAt cannot be before startsAt" });
      return;
    }

    const info = getDb()
      .prepare<[string, string, string, string | null, string]>(
        `INSERT INTO sprints (name, starts_at, ends_at, goal, status)
         VALUES (?, ?, ?, ?, ?)`
      )
      .run(name, startsAt, endsAt, goal, status);
    const id = Number(info.lastInsertRowid);

    const row = getDb()
      .prepare<[number]>(`SELECT ${SELECT_COLUMNS} FROM sprints WHERE id = ?`)
      .get(id) as SprintRow;

    res.status(201).json({ sprint: shapeSprint(row) });
  });

  router.get("/sprints/:id", (req: Request, res: Response) => {
    const id = Number.parseInt(req.params["id"] ?? "", 10);
    if (!Number.isFinite(id)) {
      res.status(400).json({ error: "bad id" });
      return;
    }
    const row = getDb()
      .prepare<[number]>(`SELECT ${SELECT_COLUMNS} FROM sprints WHERE id = ?`)
      .get(id) as SprintRow | undefined;
    if (!row) {
      res.status(404).json({ error: "no such sprint" });
      return;
    }
    const cards = (
      getDb()
        .prepare<[number]>(
          `SELECT sprint_id, card_id, planned_points FROM sprint_cards WHERE sprint_id = ?`
        )
        .all(id) as SprintCardRow[]
    ).map((c) => ({
      sprintId: c.sprint_id,
      cardId: c.card_id,
      plannedPoints: c.planned_points,
    }));
    res.json({ sprint: shapeSprint(row), cards });
  });

  router.patch("/sprints/:id", (req: Request, res: Response) => {
    const id = Number.parseInt(req.params["id"] ?? "", 10);
    if (!Number.isFinite(id)) {
      res.status(400).json({ error: "bad id" });
      return;
    }
    const existing = getDb()
      .prepare<[number]>(`SELECT ${SELECT_COLUMNS} FROM sprints WHERE id = ?`)
      .get(id) as SprintRow | undefined;
    if (!existing) {
      res.status(404).json({ error: "no such sprint" });
      return;
    }

    const body = (req.body ?? {}) as {
      name?: unknown;
      startsAt?: unknown;
      endsAt?: unknown;
      goal?: unknown;
      status?: unknown;
      pointsTarget?: unknown;
      dollarTarget?: unknown;
      reviewHoursTarget?: unknown;
      archivedAt?: unknown;
    };

    // Build the SET clause from whichever fields the caller sent. Each
    // accepted field undergoes one of three validations:
    //   - typeof check
    //   - enum membership (status)
    //   - finite-number coercion (targets)
    const sets: string[] = [];
    const params: Array<string | number | null> = [];

    if (typeof body.name === "string") {
      const v = body.name.trim();
      if (!v) {
        res.status(400).json({ error: "name cannot be empty" });
        return;
      }
      sets.push("name = ?");
      params.push(v);
    }
    if (typeof body.startsAt === "string") {
      sets.push("starts_at = ?");
      params.push(body.startsAt);
    }
    if (typeof body.endsAt === "string") {
      sets.push("ends_at = ?");
      params.push(body.endsAt);
    }
    if (body.goal === null || typeof body.goal === "string") {
      sets.push("goal = ?");
      params.push(body.goal === null ? null : (body.goal as string));
    }
    if (typeof body.status === "string") {
      if (!VALID_STATUSES.has(body.status)) {
        res.status(400).json({
          error: "status must be one of",
          valid: Array.from(VALID_STATUSES),
        });
        return;
      }
      sets.push("status = ?");
      params.push(body.status);
    }
    if (body.pointsTarget === null || typeof body.pointsTarget === "number") {
      sets.push("points_target = ?");
      params.push(
        body.pointsTarget === null ? null : Math.max(0, Math.floor(body.pointsTarget))
      );
    }
    if (body.dollarTarget === null || typeof body.dollarTarget === "number") {
      sets.push("dollar_target = ?");
      params.push(
        body.dollarTarget === null ? null : Math.max(0, body.dollarTarget)
      );
    }
    if (
      body.reviewHoursTarget === null ||
      typeof body.reviewHoursTarget === "number"
    ) {
      sets.push("review_hours_target = ?");
      params.push(
        body.reviewHoursTarget === null
          ? null
          : Math.max(0, body.reviewHoursTarget)
      );
    }
    if (body.archivedAt === null || typeof body.archivedAt === "string") {
      sets.push("archived_at = ?");
      params.push(body.archivedAt === null ? null : (body.archivedAt as string));
    }

    if (sets.length === 0) {
      res.status(400).json({ error: "no recognized fields in body" });
      return;
    }

    params.push(id);
    getDb()
      .prepare(`UPDATE sprints SET ${sets.join(", ")} WHERE id = ?`)
      .run(...params);

    const row = getDb()
      .prepare<[number]>(`SELECT ${SELECT_COLUMNS} FROM sprints WHERE id = ?`)
      .get(id) as SprintRow;
    res.json({ sprint: shapeSprint(row) });
  });

  router.post("/sprints/:id/cards", (req: Request, res: Response) => {
    const id = Number.parseInt(req.params["id"] ?? "", 10);
    if (!Number.isFinite(id)) {
      res.status(400).json({ error: "bad sprint id" });
      return;
    }
    const body = req.body as
      | { cardId?: unknown; plannedPoints?: unknown }
      | undefined;
    const cardId = typeof body?.cardId === "string" ? body.cardId : "";
    const plannedPoints =
      typeof body?.plannedPoints === "number"
        ? Math.max(0, Math.floor(body.plannedPoints))
        : null;
    if (!cardId) {
      res.status(400).json({ error: "cardId required" });
      return;
    }

    const sprint = getDb()
      .prepare<[number]>(`SELECT id FROM sprints WHERE id = ?`)
      .get(id);
    if (!sprint) {
      res.status(404).json({ error: "no such sprint" });
      return;
    }

    getDb()
      .prepare<[number, string, number | null]>(
        `INSERT OR REPLACE INTO sprint_cards (sprint_id, card_id, planned_points)
         VALUES (?, ?, ?)`
      )
      .run(id, cardId, plannedPoints);

    res.status(204).end();
  });

  router.delete(
    "/sprints/:id/cards/:cardId",
    (req: Request, res: Response) => {
      const id = Number.parseInt(req.params["id"] ?? "", 10);
      const cardId = req.params["cardId"];
      if (!Number.isFinite(id) || !cardId) {
        res.status(400).json({ error: "bad id or cardId" });
        return;
      }
      getDb()
        .prepare<[number, string]>(
          `DELETE FROM sprint_cards WHERE sprint_id = ? AND card_id = ?`
        )
        .run(id, cardId);
      res.status(204).end();
    }
  );

  return router;
}

File: backend/cards_api/routers/sprints.py
"""Sprints (+ retros — same domain family). Contract §Sprints / §Retros."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Body, Depends
from sqlalchemy.orm import Session, sessionmaker

from .. import repository as repo
from .. import wire
from ..auth import ParadigmClaims
from ..bus import OrgEventBus
from ..deps import require_claims
from ..models import SPRINT_STATUSES
from .common import err, get_bus, get_db, mutation, reading, require_json_object

router = APIRouter()


@router.get("/api/sprints")
def list_sprints(
    includeArchived: str | None = None,  # noqa: N803 - legacy query param casing
    claims: ParadigmClaims = Depends(require_claims),
    factory: sessionmaker[Session] = Depends(get_db),
) -> dict:
    include = (includeArchived or "").lower() in ("1", "true", "yes")
    with reading(factory, claims) as session:
        rows = repo.list_sprints(session, claims.org_id, include_archived=include)
        return {
            "sprints": [
                {**wire.sprint_shape(s), "cardCount": count, "plannedPointsSum": points}
                for s, count, points in rows
            ]
        }


@router.post("/api/sprints", status_code=201)
def create_sprint(
    body: dict[str, Any] = Body(...),
    claims: ParadigmClaims = Depends(require_claims),
    factory: sessionmaker[Session] = Depends(get_db),
    bus: OrgEventBus = Depends(get_bus),
) -> dict:
    body = require_json_object(body)
    name = body.get("name")
    if not isinstance(name, str) or not name.strip():
        raise err(400, "name is required")
    starts_at, ends_at = body.get("startsAt"), body.get("endsAt")
    if not isinstance(starts_at, str) or not starts_at or not isinstance(ends_at, str) or not ends_at:
        raise err(400, "startsAt and endsAt are required (ISO date)")
    if ends_at < starts_at:
        raise err(400, "endsAt cannot be before startsAt")
    status = body.get("status")
    if not isinstance(status, str) or status not in SPRINT_STATUSES:
        status = "planning"
    goal = body.get("goal")
    goal = goal if isinstance(goal, str) and goal.strip() else None

    with mutation(factory, claims, bus) as ctx:
        sprint = repo.create_sprint(
            ctx.session,
            claims.org_id,
            name=name.strip(),
            starts_at=starts_at,
            ends_at=ends_at,
            goal=goal,
            status=status,
        )
        ctx.audit("sprint.create", resource_type="sprint", resource_id=str(sprint.id))
        response = {"sprint": wire.sprint_shape(sprint)}
    return response


@router.get("/api/sprints/{sprint_id}")
def get_sprint(
    sprint_id: int,
    claims: ParadigmClaims = Depends(require_claims),
    factory: sessionmaker[Session] = Depends(get_db),
) -> dict:
    with reading(factory, claims) as session:
        sprint = repo.get_sprint(session, claims.org_id, sprint_id)
        if sprint is None:
            raise err(404, "no such sprint")
        cards = repo.sprint_cards(session, claims.org_id, sprint_id)
        return {
            "sprint": wire.sprint_shape(sprint),
            "cards": [wire.sprint_card_link(c) for c in cards],
        }


@router.patch("/api/sprints/{sprint_id}")
def patch_sprint(
    sprint_id: int,
    body: dict[str, Any] = Body(...),
    claims: ParadigmClaims = Depends(require_claims),
    factory: sessionmaker[Session] = Depends(get_db),
    bus: OrgEventBus = Depends(get_bus),
) -> dict:
    body = require_json_object(body)
    changes: dict[str, Any] = {}

    if "name" in body:
        name = body["name"]
        if isinstance(name, str):
            name = name.strip()
            if not name:
                raise err(400, "name cannot be empty")
            changes["name"] = name
    for wire_key, column in (("startsAt", "starts_at"), ("endsAt", "ends_at")):
        if wire_key in body and isinstance(body[wire_key], str) and body[wire_key]:
            changes[column] = body[wire_key]
    for wire_key, column in (("goal", "goal"), ("archivedAt", "archived_at")):
        if wire_key in body and (body[wire_key] is None or isinstance(body[wire_key], str)):
            changes[column] = body[wire_key]
    if "status" in body:
        status = body["status"]
        if isinstance(status, str):
            if status not in SPRINT_STATUSES:
                raise err(400, "status must be one of", valid=list(SPRINT_STATUSES))
            changes["status"] = status
    if "pointsTarget" in body:
        value = body["pointsTarget"]
        if value is None:
            changes["points_target"] = None
        elif isinstance(value, (int, float)) and not isinstance(value, bool):
            changes["points_target"] = max(0, int(value))
    for wire_key, column in (
        ("dollarTarget", "dollar_target"),
        ("reviewHoursTarget", "review_hours_target"),
    ):
        if wire_key in body:
            value = body[wire_key]
            if value is None:
                changes[column] = None
            elif isinstance(value, (int, float)) and not isinstance(value, bool):
                changes[column] = max(0.0, float(value))

    if not changes:
        raise err(400, "no recognized fields in body")

    with mutation(factory, claims, bus) as ctx:
        sprint = repo.get_sprint(ctx.session, claims.org_id, sprint_id)
        if sprint is None:
            raise err(404, "no such sprint")
        for column, value in changes.items():
            setattr(sprint, column, value)
        ctx.audit("sprint.update", resource_type="sprint", resource_id=str(sprint_id))
        response = {"sprint": wire.sprint_shape(sprint)}
    return response


@router.post("/api/sprints/{sprint_id}/cards", status_code=204)
def add_sprint_card(
    sprint_id: int,
    body: dict[str, Any] = Body(...),
    claims: ParadigmClaims = Depends(require_claims),
    factory: sessionmaker[Session] = Depends(get_db),
    bus: OrgEventBus = Depends(get_bus),
) -> None:
    body = require_json_object(body)
    card_id = body.get("cardId")
    if not isinstance(card_id, str) or not card_id:
        raise err(400, "cardId required")
    planned = body.get("plannedPoints")
    planned_points = (
        max(0, int(planned))
        if isinstance(planned, (int, float)) and not isinstance(planned, bool)
        else None
    )
    with mutation(factory, claims, bus) as ctx:
        if repo.get_sprint(ctx.session, claims.org_id, sprint_id) is None:
            raise err(404, "no such sprint")
        repo.upsert_sprint_card(ctx.session, claims.org_id, sprint_id, card_id, planned_points)
        ctx.audit(
            "sprint.add_card",
            resource_type="sprint",
            resource_id=str(sprint_id),
            details={"cardId": card_id},
        )


@router.delete("/api/sprints/{sprint_id}/cards/{card_id}", status_code=204)
def remove_sprint_card(
    sprint_id: int,
    card_id: str,
    claims: ParadigmClaims = Depends(require_claims),
    factory: sessionmaker[Session] = Depends(get_db),
    bus: OrgEventBus = Depends(get_bus),
) -> None:
    with mutation(factory, claims, bus) as ctx:
        repo.remove_sprint_card(ctx.session, claims.org_id, sprint_id, card_id)
        ctx.audit(
            "sprint.remove_card",
            resource_type="sprint",
            resource_id=str(sprint_id),
            details={"cardId": card_id},
        )


# --------------------------------------------------------------------------
# Retros (wire shape is raw snake_case — preserved legacy quirk)


@router.get("/api/retros")
def list_retros(
    claims: ParadigmClaims = Depends(require_claims),
    factory: sessionmaker[Session] = Depends(get_db),
) -> dict:
    with reading(factory, claims) as session:
        retros = repo.list_retros(session, claims.org_id)
        return {"retros": [wire.retro_row(r) for r in retros]}


@router.post("/api/retros", status_code=201)
def create_retro(
    body: dict[str, Any] = Body(...),
    claims: ParadigmClaims = Depends(require_claims),
    factory: sessionmaker[Session] = Depends(get_db),
    bus: OrgEventBus = Depends(get_bus),
) -> dict:
    body = require_json_object(body)
    held_on = body.get("heldOn")
    if not isinstance(held_on, str) or not held_on:
        raise err(400, "heldOn required (ISO date)")
    sprint_id = body.get("sprintId")
    sprint_id = (
        int(sprint_id)
        if isinstance(sprint_id, (int, float)) and not isinstance(sprint_id, bool)
        else None
    )
    summary = body.get("summary")
    summary = summary if isinstance(summary, str) else None
    with mutation(factory, claims, bus) as ctx:
        retro = repo.create_retro(
            ctx.session, claims.org_id, sprint_id=sprint_id, held_on=held_on, summary=summary
        )
        ctx.audit("retro.create", resource_type="retro", resource_id=str(retro.id))
        response = {"id": retro.id}
    return response


@router.get("/api/retros/{retro_id}")
def get_retro(
    retro_id: int,
    claims: ParadigmClaims = Depends(require_claims),
    factory: sessionmaker[Session] = Depends(get_db),
) -> dict:
    with reading(factory, claims) as session:
        retro = repo.get_retro(session, claims.org_id, retro_id)
        if retro is None:
            raise err(404, "no such retro")
        return wire.retro_row(retro)
